patches/api/0469-Adding-PredicateChoice.patch

```diff
@@ -0,0 +1,94 @@
++import java.util.List;
++import java.util.function.Predicate;
++
++/**
++ * Package private implementation for {@link org.bukkit.inventory.RecipeChoice.PredicateChoice}
++ * @param predicate - The Item predicate
++ * @param choices - The recipe book choices
++ */
++record PredicateChoiceImpl(Predicate<ItemStack> predicate, List<ItemStack> choices) implements RecipeChoice.PredicateChoice {
++    @Override
++    public @NotNull ItemStack getItemStack() {
++        ItemStack stack = new ItemStack(choices.get(0));
++        // For compat
++        if (choices.size() > 1) {
++            stack.setDurability(Short.MAX_VALUE);
++            return stack;
++        }
++        return stack;
++    }
++
++    @Override
++    public @NotNull RecipeChoice clone() {
++        return new PredicateChoiceImpl(predicate, recipeBookExamples());
++    }
++
++    @Override
++    public boolean test(@NotNull final ItemStack itemStack) {
++        return predicate.test(itemStack);
++    }
++
++    @Override
++    public List<ItemStack> recipeBookExamples() {
++        return List.copyOf(choices);
++    }
++}
+diff --git a/src/main/java/org/bukkit/inventory/RecipeChoice.java b/src/main/java/org/bukkit/inventory/RecipeChoice.java
+index db8bcc66bdc4bedfffb4705db6338eda4c0ad29a..e331040030c484ad63444a0ee4e42790d639115e 100644
+--- a/src/main/java/org/bukkit/inventory/RecipeChoice.java
++++ b/src/main/java/org/bukkit/inventory/RecipeChoice.java
+@@ -233,4 +233,39 @@ public interface RecipeChoice extends Predicate<ItemStack>, Cloneable {
+             return "ExactChoice{" + "choices=" + choices + '}';
+         }
+     }
++    // Paper start - Adding PredicateChoice
++    /**
++     * Represents a choice that matches when the item predicate is fulfilled.
++     */
++
++    interface PredicateChoice extends RecipeChoice {
++        static PredicateChoice create(@NotNull Predicate<ItemStack> predicate, ItemStack... recipeBookExamples){
++            Objects.requireNonNull(predicate, "The item predicate cannot be null!");
++            Objects.requireNonNull(predicate, "The mustHaveRecipeBookExample cannot be null!");
++            if(recipeBookExamples.length == 0)
++                throw new IllegalArgumentException("Please provide at least one recipe book example item!");
++            return new PredicateChoiceImpl(predicate, List.of(recipeBookExamples));
++        }
++
++        static PredicateChoice create(@NotNull Predicate<ItemStack> predicate, java.util.Collection<ItemStack> recipeBookExamples){
++            Objects.requireNonNull(predicate, "The item predicate cannot be null!");
++            Objects.requireNonNull(predicate, "The mustHaveRecipeBookExample cannot be null!");
++            if(recipeBookExamples.isEmpty())
++                throw new IllegalArgumentException("Please provide at least one recipe book example item!");
++            return new PredicateChoiceImpl(predicate, List.copyOf(recipeBookExamples));
++        }
```

Comment:
I would move these methods up 1 level to be on RecipeChoice itself and named `createPredicateChoice`

An alternative would be to just make PredicateChoice a top-level class next to RecipeChoice and have `create` methods in it. Not a fan of the static factory methods inside an inner class.